WatchFaceAlphaKotlin/app/src/main/java/com/example/android/wearable/alpha/AnalogWatchFaceService.kt

```diff
@@ -0,0 +1,322 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.example.android.wearable.alpha
+
+import android.content.Context
+import android.graphics.RectF
+import android.graphics.drawable.Icon
+import android.util.Log
+import android.view.SurfaceHolder
+import androidx.wear.complications.ComplicationBounds
+import androidx.wear.complications.DefaultComplicationProviderPolicy
+import androidx.wear.complications.SystemProviders
+import androidx.wear.complications.data.ComplicationType
+import androidx.wear.watchface.CanvasComplicationDrawable
+import androidx.wear.watchface.CanvasType
+import androidx.wear.watchface.Complication
+import androidx.wear.watchface.ComplicationsManager
+import androidx.wear.watchface.WatchFace
+import androidx.wear.watchface.WatchFaceService
+import androidx.wear.watchface.WatchFaceType
+import androidx.wear.watchface.WatchState
+import androidx.wear.watchface.complications.rendering.ComplicationDrawable
+import androidx.wear.watchface.style.Layer
+import androidx.wear.watchface.style.UserStyleRepository
+import androidx.wear.watchface.style.UserStyleSchema
+import androidx.wear.watchface.style.UserStyleSetting
+import com.example.android.wearable.alpha.data.db.WatchFaceColorStyleEntity
+import com.example.android.wearable.alpha.viewmodels.AnalogWatchFaceViewModel
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.MainScope
+import kotlinx.coroutines.cancel
+import kotlinx.coroutines.launch
+
+/**
+ * Handles much of the boilerplate needed to implement a watch face (minus rendering code; see
+ * [AnalogWatchCanvasRenderer]) including the complications and settings (styles user can change on
+ * the watch face).
+ */
+class AnalogWatchFaceService : WatchFaceService() {
+    // Used to launch coroutines (non-blocking way to pull watch face color styles data [only
+    // needed on the initial load of the user color style options]).
+    private val scope: CoroutineScope = MainScope()
+
+    private lateinit var analogWatchFaceViewModel: AnalogWatchFaceViewModel
+
+    private var colorStylesList: List<WatchFaceColorStyleEntity>? = null
+    private var complicationStyleDrawableId = R.drawable.complication_white_style
+
+    override fun onCreate() {
+        super.onCreate()
+
+        analogWatchFaceViewModel =
+            AnalogWatchFaceViewModel((application as MainApplication).repository)
+
+        scope.launch {
+            colorStylesList =
+                analogWatchFaceViewModel.getAllWatchFaceColorStyles()
+        }
+    }
+
+    override fun onDestroy() {
+        analogWatchFaceViewModel.clear()
+        // Cancels scope used for retrieving watch face styles.
+        scope.cancel("AnalogWatchFaceService.onDestroy()")
+        super.onDestroy()
+    }
+
+    override fun createWatchFace(surfaceHolder: SurfaceHolder, watchState: WatchState): WatchFace {
```

Comment:
I would suggest breaking up the configuration code from the boilerplate. For example, the `createUserStyles` method at this point not only defines the actual styles to use (configuration) but also takes on the boilerplate of creating a repository based on that. It would be great if a developer can clone this repo and have some very specific classes to adapt to their use case. Right now they would have to search through the code to see which parts need to be changed.

Reply:
I broke these out to an extension function and several utility functions in the util directory.

PTAL and let me know if that makes more sense.